### PR TITLE
Tpetra: Add LinearProblem

### PR DESCRIPTION
**Pull request by ccober6**
As part of the effort to transition from Epetra to Tpetra, several features of EpetraExt need transition (e.g., Singleton Filtering). To begin, we need to create a Tpetra version of LinearProblem, which includes left and right scaling.  Basic unit testing is also included.

@trilinos/tpetra 

## Motivation
Need to transition several transform functions from Epetra to Tpetra, and LinearProblem underlies these functions.

## Related Issues
None

## Stakeholder Feedback
We are on contact with the requesting application, but needed functionality is yet to come.

## Testing
Basic unit testing is included.

### CODE REVIEW

File: packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp
Comment (jhux2) [inline packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp:151]: @ccober6 I'm not a fan of returning error codes.  What about just throwing an exception or a hard failure?

**Review by jhux2 — COMMENTED** [review]
See my comment about error codes.

File: packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp
Comment (ccober6) [reply packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp:151]: Yeah, same here and I initially started that way, but did not know what Tpetra's "policy" was, and what the stakeholder usage looked like to maintain similar functionality.  That's why I optionally added it.  I am open to removing the error codes.

Comment (jhux2) [reply packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp:151]: There's no backwards compatibility to worry about because this is new functionality.  I'll let others chime in, but I think throwing an exception is probably the way to go.

Comment (ccober6) [reply packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp:151]: Yeah, not true backward compatibility but work the same as Epetra_LinearProblem::checkInput().  Throwing is probably the best route.

Comment (ccober6) [reply packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp]: @jhux2, ok removed the error code.  Can I get another review?

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (jhux2) [inline packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:245]: Use Teuchos `eps` instead?

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (jhux2) [inline packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:247]: Use Teuchos `eps` instead?

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (jhux2) [inline packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:32]: Can all the display utilities be commented or macro'd out, since their use is also commented out below?

**Review by jhux2 — CHANGES_REQUESTED** [review]

File: packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp
Comment (ccober6) [reply packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp:52]: Good idea.

File: packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp
Comment (ccober6) [reply packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp:67]: Good idea.

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (ccober6) [reply packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:174]: Done.

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (ccober6) [reply packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:235]: Done.

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (ccober6) [reply packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:245]: Done.

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (ccober6) [reply packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:247]: Done.

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (ccober6) [reply packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:243]: Done.

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (ccober6) [reply packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:32]: I added a macro.

**Comment by ccober6** [comment]
@jhux2, can I get hopefully one last review on this?  :)

File: packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp
Comment (jhux2) [inline packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp:126]: Minor quibble that this isn't tested.  I won't let this block approval, though.

File: packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp
Comment (jhux2) [inline packages/tpetra/core/test/LinearProblem/LinearProblem_UnitTests.cpp:217]: `linearProblem->checkInput()` could be called before setting A, X, and B, respectively, to test the error checking.

**Review by jhux2 — APPROVED** [review]
Lgtm.  Two suggestions for future improvements to testing.

**Pull request merged**
